# Lecture 09: Relational Data
<div style="border: 1px double black; padding: 10px; margin: 10px">

**Goals for today's lecture:**
* Learn about [keys](#Keys)
* Different types of [relations](#Relations) between tables
* [Commands for joining related tables together](#Outer-joins)
</div>


We have already spent a lot of time analyzing the `flights` table. In fact, there are four other tables in `nycflights13` that contain related information about these flights:

In [2]:
install.packages('nycflights13')
library(tidyverse)
library(nycflights13)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.3     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


In [ ]:
print(airlines)

# A tibble: 16 × 2
   carrier name                       
   <chr>   <chr>                      
 1 9E      Endeavor Air Inc.          
 2 AA      American Airlines Inc.     
 3 AS      Alaska Airlines Inc.       
 4 B6      JetBlue Airways            
 5 DL      Delta Air Lines Inc.       
 6 EV      ExpressJet Airlines Inc.   
 7 F9      Frontier Airlines Inc.     
 8 FL      AirTran Airways Corporation
 9 HA      Hawaiian Airlines Inc.     
10 MQ      Envoy Air                  
11 OO      SkyWest Airlines Inc.      
12 UA      United Air Lines Inc.      
13 US      US Airways Inc.            
14 VX      Virgin America             
15 WN      Southwest Airlines Co.     
16 YV      Mesa Airlines Inc.         


In [ ]:
airports %>% filter(faa == "DTW")

  faa name                   lat      lon       alt tz dst tzone           
1 DTW Detroit Metro Wayne Co 42.21244 -83.35339 645 -5 A   America/New_York

In [ ]:
print(planes)

# A tibble: 3,322 x 9
   tailnum  year type          manufacturer   model  engines seats speed engine 
   <chr>   <int> <chr>         <chr>          <chr>    <int> <int> <int> <chr>  
 1 N10156   2004 Fixed wing m… EMBRAER        EMB-1…       2    55    NA Turbo-…
 2 N102UW   1998 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 3 N103US   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 4 N104UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 5 N10575   2002 Fixed wing m… EMBRAER        EMB-1…       2    55    NA Turbo-…
 6 N105UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 7 N107US   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 8 N108UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 9 N109UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
10 N110UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
# … wi

In [ ]:
print(weather)

# A tibble: 26,115 x 15
   origin  year month   day  hour  temp  dewp humid wind_dir wind_speed
   <chr>  <dbl> <dbl> <int> <int> <dbl> <dbl> <dbl>    <dbl>      <dbl>
 1 EWR     2013     1     1     1  39.0  26.1  59.4      270      10.4 
 2 EWR     2013     1     1     2  39.0  27.0  61.6      250       8.06
 3 EWR     2013     1     1     3  39.0  28.0  64.4      240      11.5 
 4 EWR     2013     1     1     4  39.9  28.0  62.2      250      12.7 
 5 EWR     2013     1     1     5  39.0  28.0  64.4      260      12.7 
 6 EWR     2013     1     1     6  37.9  28.0  67.2      240      11.5 
 7 EWR     2013     1     1     7  39.0  28.0  64.4      240      15.0 
 8 EWR     2013     1     1     8  39.9  28.0  62.2      250      10.4 
 9 EWR     2013     1     1     9  39.9  28.0  62.2      260      15.0 
10 EWR     2013     1     1    10  41    28.0  59.6      260      13.8 
# … with 26,105 more rows, and 5 more variables: wind_gust <dbl>, precip <dbl>,
#   pressure <dbl>, visib <dbl>,

In [ ]:
weather$origin %>% unique

[1] "EWR" "JFK" "LGA"

Together these four tables form a *relational database*. The relationships can be graphed like so:
![table relationships](http://r4ds.had.co.nz/diagrams/relational-nycflights.png)

The particular relationships in this database are:
- `flights` connects to `planes` via `tailnum`.
- `flights` connects to `airlines` via `carrier`.
- `flights` connects to `airports` twice: via `origin` and `dest`.
- `flights` connects to `weather` via `origin` (the location), and `year`, `month`, `day` and `hour`.

## Keys
The "key" to understanding relational databases is... keys.



### Primary Key
A *primary key* is a variable (or set of variables) that uniquely identifies an observation in its own table: there is **at most** one row in the table that corresponds to any setting of the columns which comprise the key.

A primary key is
* a variable or
* set of variables that uniquely identify each observation (compound key).

When more than one variable is needed, the key is called a compound key.

In the `planes` table, each airplane is identified by its `tailnum`:

In [ ]:
print(planes)

# A tibble: 3,322 x 9
   tailnum  year type          manufacturer   model  engines seats speed engine 
   <chr>   <int> <chr>         <chr>          <chr>    <int> <int> <int> <chr>  
 1 N10156   2004 Fixed wing m… EMBRAER        EMB-1…       2    55    NA Turbo-…
 2 N102UW   1998 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 3 N103US   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 4 N104UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 5 N10575   2002 Fixed wing m… EMBRAER        EMB-1…       2    55    NA Turbo-…
 6 N105UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 7 N107US   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 8 N108UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
 9 N109UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
10 N110UW   1999 Fixed wing m… AIRBUS INDUST… A320-…       2   182    NA Turbo-…
# … wi

The tail number of an airplane is assigned by a government agency and is unique: no two planes can have the same tail number. Thus, `tailnum` should be a primary key in this table.

To check that one or more variables constitutes a primary key, we can group by those variables and then check that the number of distinct values equals the number of rows in the data set:

In [31]:
# planes %>% print
planes %>% summarize(n = n(), nd = n_distinct(tailnum))

n    nd  
1 3322 3322

In [33]:
planes %>% count(tailnum) %>% filter(n > 1)

tailnum n

Compare with `flights`, where tailnum does *not* uniquely identify each row. (There are many flights present for the same airplane.)

In [ ]:
flights %>% count(tailnum) %>% filter(n > 1) %>% print

# A tibble: 3,873 × 2
   tailnum     n
   <chr>   <int>
 1 D942DN      4
 2 N0EGMQ    371
 3 N10156    153
 4 N102UW     48
 5 N103US     46
 6 N104UW     47
 7 N10575    289
 8 N105UW     45
 9 N107US     41
10 N108UW     60
# ℹ 3,863 more rows


What is the primary key for the `flights` table?

In [34]:
flights %>% count(year, month, day, dep_time, tailnum) %>% filter (n > 1)

year month day dep_time tailnum n 
1   2013 1      2  NA       N10575   2
2   2013 1      2  NA       NA       2
3   2013 1      3  NA       NA       2
4   2013 1      4  NA       NA       2
5   2013 1      9  NA       NA       2
6   2013 1     10  NA       NA       2
7   2013 1     12  NA       NA       2
8   2013 1     13  NA       NA       7
9   2013 1     15  NA       NA       2
10  2013 1     16  NA       N12195   2
11  2013 1     16  NA       NA      24
12  2013 1     17  NA       NA       4
13  2013 1     18  NA       N12166   2
14  2013 1     18  NA       N12957   2
15  2013 1     18  NA       NA       3
16  2013 1     21  NA       NA       5
17  2013 1     22  NA       NA       3
18  2013 1     23  NA       NA       4
19  2013 1     24  NA       NA       4
20  2013 1     25  NA       N13133   2
21  2013 1     25  NA       N13968   2
22  2013 1     25  NA       NA      16
23  2013 1     26  NA       N832MQ   2
24  2013 1     26  NA       NA       2
25  2013 1     27  NA       NA       6
26  2013 1     28  NA       N10156   2
27  2013 1     28  NA       N11548   2
28  2013 1     28  NA       N13908   2
29  2013 1     28  NA       N13969   3
30  2013 1     28  NA       N14125   2
⋮   ⋮    ⋮     ⋮   ⋮        ⋮       ⋮ 
677 2013 12    14  NA       N13992   2
678 2013 12    14  NA       N14542   2
679 2013 12    14  NA       N16951   2
680 2013 12    14  NA       N16987   2
681 2013 12    14  NA       N31131   3
682 2013 12    14  NA       N539MQ   2
683 2013 12    14  NA       N829AS   2
684 2013 12    14  NA       NA      38
685 2013 12    15  NA       N13989   2
686 2013 12    15  NA       N661JB   2
687 2013 12    15  NA       NA      21
688 2013 12    16  NA       N816MQ   2
689 2013 12    17  NA       N13566   2
690 2013 12    17  NA       N13949   2
691 2013 12    17  NA       N14186   2
692 2013 12    17  NA       N14570   2
693 2013 12    17  NA       N14904   2
694 2013 12    17  NA       N16918   2
695 2013 12    17  NA       N16981   2
696 2013 12    17  NA       N813MQ   2
697 2013 12    17  NA       NA      27
698 2013 12    20  NA       N16112   2
699 2013 12    20  NA       NA       3
700 2013 12    21  NA       NA       2
701 2013 12    23  NA       NA      10
702 2013 12    26  NA       N877AS   2
703 2013 12    29  NA       N231JB   2
704 2013 12    29  NA       NA       2
705 2013 12    30  NA       NA       4
706 2013 12    31  NA       NA      11

We might guess that `year`, `month`, `day`, and `tailnum` are sufficient to identify each row in `flights`, but this is not true:

In [35]:
flights %>% summarize(n = n(), nd = n_distinct(year, month, day, tailnum))

n      nd    
1 336776 251727

In fact, even restricting to the exact *minute* that an airplane departed is not sufficient:

In [36]:
flights %>% summarize(n=n(), nd=n_distinct(tailnum, time_hour, minute))

n      nd    
1 336776 336367

In [37]:
flights %>% summarize(n=n(), nd = n_distinct(tailnum, year, month, day, hour, minute))

n      nd    
1 336776 336367

This says that there are certain airplanes that are marked as having departed more than once in the same year, month, day, hour and minute. We can inspect these rows as follows:

In [ ]:
count(flights, tailnum, time_hour, minute) %>% filter(n > 1) %>% print

# A tibble: 298 × 4
   tailnum time_hour           minute     n
   <chr>   <dttm>               <dbl> <int>
 1 N11119  2013-06-10 16:00:00     55     2
 2 N11192  2013-08-26 08:00:00     30     2
 3 N12563  2013-02-04 16:00:00     19     2
 4 N12564  2013-01-13 20:00:00      0     2
 5 N12900  2013-07-10 21:00:00     29     2
 6 N13969  2013-01-28 07:00:00     59     2
 7 N14148  2013-03-12 06:00:00     30     2
 8 N14558  2013-04-19 13:00:00     29     2
 9 N14916  2013-02-11 13:00:00     15     2
10 N14974  2013-07-26 06:00:00     30     2
# ℹ 288 more rows


In [ ]:
count(flights, tailnum, year, month, day, hour, minute) %>% filter(n > 1) %>% print

# A tibble: 298 × 7
   tailnum  year month   day  hour minute     n
   <chr>   <int> <int> <int> <dbl>  <dbl> <int>
 1 N11119   2013     6    10    16     55     2
 2 N11192   2013     8    26     8     30     2
 3 N12563   2013     2     4    16     19     2
 4 N12564   2013     1    13    20      0     2
 5 N12900   2013     7    10    21     29     2
 6 N13969   2013     1    28     7     59     2
 7 N14148   2013     3    12     6     30     2
 8 N14558   2013     4    19    13     29     2
 9 N14916   2013     2    11    13     15     2
10 N14974   2013     7    26     6     30     2
# ℹ 288 more rows


These likely indicate data entry errors.

## 🤔 Quiz

What column(s) constitute a primary key in the `mpg` table?

<ol style="list-style-type: upper-alpha;">
    <li><code>manufacturer</code> and <code>model</code></li>
    <li><code>manufacturer</code>, <code>year</code>, and <code>model</code></li>
    <li><code>manufacturer</code>, <code>year</code>, <code>displ</code>, and <code>model</code></li>
    <li><code>manufacturer</code>, <code>year</code>, <code>displ</code>, <code>trans</code>, and <code>model</code></li>
    <li>None of these</li>
</ol>



In [42]:
# primary key in mpg
mpg %>% count(manufacturer, model, year, displ, trans) %>% filter( n > 1)

manufacturer model               year displ trans      n
1  chevrolet    c1500 suburban 2wd  2008 5.3   auto(l4)   3
2  chevrolet    k1500 tahoe 4wd     2008 5.3   auto(l4)   2
3  dodge        caravan 2wd         1999 3.3   auto(l4)   2
4  dodge        caravan 2wd         1999 3.8   auto(l4)   2
5  dodge        caravan 2wd         2008 3.3   auto(l4)   3
6  dodge        dakota pickup 4wd   2008 4.7   auto(l5)   3
7  dodge        durango 4wd         2008 4.7   auto(l5)   3
8  dodge        ram 1500 pickup 4wd 2008 4.7   auto(l5)   3
9  dodge        ram 1500 pickup 4wd 2008 4.7   manual(m6) 3
10 ford         explorer 4wd        1999 4.0   auto(l5)   2
11 honda        civic               1999 1.6   auto(l4)   2
12 honda        civic               1999 1.6   manual(m5) 3
13 honda        civic               2008 1.8   auto(l5)   2
14 jeep         grand cherokee 4wd  2008 4.7   auto(l5)   2
15 lincoln      navigator 2wd       1999 5.4   auto(l4)   2
16 pontiac      grand prix          1999 3.8   auto(l4)   2
17 subaru       forester awd        2008 2.5   auto(l4)   2
18 subaru       forester awd        2008 2.5   manual(m5) 2
19 subaru       impreza awd         2008 2.5   auto(s4)   2
20 subaru       impreza awd         2008 2.5   manual(m5) 2

## Relations
A foreign key is a variable (or set of variables) that corresponds to a primary key in another table. For example:
* `flights$tailnum` is a foreign key that corresponds to the primary key `planes$tailnum`.
* `flights$carrier` is a foreign key that corresponds to the primary key `airlines$carrier`.
* `flights$origin` is a foreign key that corresponds to the primary key `airports$faa`.
* `flights$dest` is a foreign key that corresponds to the primary key `airports$faa`.
  
A **primary key** and the corresponding **foreign key** in another table form a *relation*. Relations come in several forms:
- *One-to-many*. (Most common). For example, each flight has one plane, but each plane has many flights.
- *Many-to-many*: For example, each airline flies to many airports; each airport hosts many airlines.
- *One-to-one*. Each row in one table corresponds uniquely to a row in a second table. This is relatively uncommon because you could just as easily combine the two tables into one.

In [3]:
x = tribble(
  ~key, ~val_x,
     1, "x1",
     2, "x2",
     3, "x3"
)
y = tribble(
  ~key, ~val_y,
     1, "y1",
     2, "y2",
     4, "y3"
)

In [47]:
x
y

k val_x
1 1 x1   
2 2 x2   
3 3 x3

key val_y
1 1   y1   
2 2   y2   
3 4   y3

## Joins
Joins are the way we combine or "merge" two data tables based on keys.
To understand how joins work we'll study these two simple tables:

![simple tables](http://r4ds.had.co.nz/diagrams/join-setup.png)

### Inner joins
Inner joins match a pair of observations whenever their keys are equal:
![match example](https://r4ds.hadley.nz/diagrams/join/inner.png)

In [4]:
# older way of joining
x %>% inner_join(y, by = "key")

key val_x val_y
1 1   x1    y1   
2 2   x2    y2

In [5]:
# using join_by is preferred
x %>% inner_join(y, join_by('key'))

key val_x val_y
1 1   x1    y1   
2 2   x2    y2

In [6]:
x %>% inner_join(y, join_by('key' == 'key'))

key val_x val_y
1 1   x1    y1   
2 2   x2    y2

Note that there is no row for `key=3` or `key=4`: with an inner join, unmatched rows are not included in the result. For this reason, we do not as often use inner joins for data analysis since you can easily lose observations.

#### Differences between `by` and `join_by`
* by = "x" corresponds to join_by(x).
* by = c("a" = "x") corresponds to join_by(a == x).


### Outer joins
An outer join keeps observations that appear in at least one of the tables. There are three types of outer joins:
- A left join keeps all observations in x.
- A right join keeps all observations in y.
- A full join keeps all observations in both x and y.

![match example](https://r4ds.hadley.nz/diagrams/join/left.png)

![match example](https://r4ds.hadley.nz/diagrams/join/right.png)

![match example](https://r4ds.hadley.nz/diagrams/join/full.png)

In [57]:
x %>% full_join(y, join_by(k == key))

k val_x val_y
1 1 x1    y1   
2 2 x2    y2   
3 3 x3    NA   
4 4 NA    y3

In [53]:
x
y

k val_x
1 1 x1   
2 2 x2   
3 3 x3

key val_y
1 1   y1   
2 2   y2   
3 4   y3

Left joins are the most common. Use them to look up data in another table, but preserve your original observations when there are records on the left table for which the other table does not have a match.

### Example
The `flights` table has a `carrier` column which is a two-letter code for the airline. The `airlines` table maps these code to recognizable airline names.

How many flights are there per carrier name (full name of the carrier) in the dataset?

In [62]:
# flights per carrier

flights %>%
  count(carrier) %>%
    left_join(airlines, join_by(carrier)) %>%
      select(name, n)

name                        n    
1  Endeavor Air Inc.           18460
2  American Airlines Inc.      32729
3  Alaska Airlines Inc.          714
4  JetBlue Airways             54635
5  Delta Air Lines Inc.        48110
6  ExpressJet Airlines Inc.    54173
7  Frontier Airlines Inc.        685
8  AirTran Airways Corporation  3260
9  Hawaiian Airlines Inc.        342
10 Envoy Air                   26397
11 SkyWest Airlines Inc.          32
12 United Air Lines Inc.       58665
13 US Airways Inc.             20536
14 Virgin America               5162
15 Southwest Airlines Co.      12275
16 Mesa Airlines Inc.            601

## 🤔 Quiz

How many flights departing from LGA were operated by JetBlue Airways?

<ol style="list-style-type: upper-alpha;">
    <li>46087</li>
    <li>26397</li>
    <li>2276</li>
    <li>6002</li>
    <li>43939</li>
</ol>

In [66]:
# jetblue from LGA
flights %>% filter(origin == 'LGA') %>%
  left_join(airlines) %>%
    filter (name == 'JetBlue Airways') %>%
      nrow

Joining with `by = join_by(carrier)`


[1] 6002

### Duplicate foreign keys
A primary key is unique but not a foreign key.

##### One table has duplicate keys
This is useful when you want to add in additional information as there is typically a one-to-many relationship.
![duplicate keys](http://r4ds.had.co.nz/diagrams/join-one-to-many.png)

In fact, we already saw an example of this above when we joined `planes` to `flights`: each plane belongs to potentially many flights.

### Exercise
What is the most common model of airplane used by each carrier?

In [72]:
# most common model

flights %>%
  inner_join(planes, join_by(tailnum)) %>%
  group_by(carrier, model) %>%
    summarise(n = n()) %>% top_n(1)

`summarise()` has grouped output by 'carrier'. You can override using the
`.groups` argument.
Selecting by n


carrier model       n    
1  9E      CL-600-2D24 10580
2  AA      767-223      4257
3  AS      737-890       346
4  B6      A320-232    34063
5  DL      MD-88       10191
6  EV      EMB-145LR   28027
7  F9      A320-214      617
8  FL      717-200      2774
9  HA      A330-243      342
10 MQ      G1159B        486
11 OO      CL-600-2C10    25
12 UA      737-824     13809
13 US      A319-112     5844
14 VX      A320-214     4859
15 WN      737-7H4     10389
16 YV      CL-600-2C10   311

##### When Both tables have duplicate keys
This represents a many-to-many join and is usually an error since the key does not uniquely identify observations in either table. Joining duplicated keys results in the Cartesian product of all the possible matches:
![cartesian](http://r4ds.had.co.nz/diagrams/join-many-to-many.png)

Be careful when doing many-to-many merges. It's possible to generate huge tables by accident and crash R.

## 🤔 Quiz
To understand what the weather conditions were when each flight departed, I will join the weather table to the first 100 rows of the flights table using the command

```
flights %>% slice(1:100) %>% left_join(weather, by=c("origin", "hour"))
```

How many rows does the resulting table have?

<ol style="list-style-type: upper-alpha;">
    <li>100</li>
    <li>36344</li>
    <li>336776</li>
    <li>Zero</li>
    <li><code>NA</code></li>
</ol>

In [73]:
flights %>% slice(1:100) %>% left_join(weather, by=c("origin", "hour")) %>% nrow

Warning message in left_join(., weather, by = c("origin", "hour")):
“Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 1 of `x` matches multiple rows in `y`.
ℹ Row 8708 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  "many-to-many"` to silence this warning.”


[1] 36344

### Defining the key columns
When we do a join using `left_join()`, R take as the key whatever column names the two tables have in common:

In [ ]:
left_join(flights, planes) %>% print

Joining, by = c("year", "tailnum")


# A tibble: 336,776 x 26
    year month   day dep_time sched_dep_time dep_delay arr_time sched_arr_time
   <int> <int> <int>    <int>          <int>     <dbl>    <int>          <int>
 1  2013     1     1      517            515         2      830            819
 2  2013     1     1      533            529         4      850            830
 3  2013     1     1      542            540         2      923            850
 4  2013     1     1      544            545        -1     1004           1022
 5  2013     1     1      554            600        -6      812            837
 6  2013     1     1      554            558        -4      740            728
 7  2013     1     1      555            600        -5      913            854
 8  2013     1     1      557            600        -3      709            723
 9  2013     1     1      557            600        -3      838            846
10  2013     1     1      558            600        -2      753            745
# … with 336,766 more rows,

This is called a *natural join*. If the key column(s) are named differently in the two tables, we must specify the mapping between the two using the `by=` or `join_by` parameter.


Consider joining `airports` to `flights`:
```
> left_join(flights, airports)
Error: `by` required, because the data sources have no common variables
Traceback:

1. left_join(flights, airports)
2. left_join.tbl_df(flights, airports)
3. common_by(by, x, y)
4. common_by.NULL(by, x, y)
5. bad_args("by", "required, because the data sources have no common variables")
6. glubort(fmt_args(args), ..., .envir = .envir)
7. .abort(text)
```

This has produced an error, because airports and flights do not have any columns in common. Indeed, the three digit FAA code is called `faa` in `airports`, but appears as either `origin` or `dest` in `flights`. To fix the error, we must specify which of `origin` or `dest` should be matched:

In [ ]:
# join_by

## 🤔 Quiz

How many flights were bound to the Hawaii timezone in this dataset?

<ol style="list-style-type: upper-alpha;">
    <li>560</li>
    <li>707</li>
    <li>710</li>
    <li>659</li>
    <li>500</li>
</ol>


In [ ]:
# your query

## Filtering joins

Filtering joins allow us to filter rows on one table based on their presence or absence in another table. We've already seen some examples of this on the problem sets:

In [ ]:
dest_top6 <- count(flights, dest) %>% top_n(6)
filter(flights, dest %in% dest_top6$dest) %>% nrow

Selecting by n


[1] 94326

### Semi-join
`semi_join(x, y)` keeps all the observations in `x` that are also in `y`.
![semi join](http://r4ds.had.co.nz/diagrams/join-semi.png)

In [ ]:
# # find the total count of flights that have plane information
